src/targets/gpu/fuse_mlir.cpp

```diff
@@ -678,93 +680,140 @@ using find_mlir_standalone_dot_op         = find_mlir_standalone_op<&is_mlir_dot
+        auto axes = fused_reduce->get_operator().to_value()["axes"];
+        if(axes.size() != 1)
+            return;
+
+        module_ref m_attn =
+            mpm.create_module("mlir_attn_" + fused_reduce->module_inputs().front()->name());
+        m_attn->set_bypass();
+        std::unordered_map<instruction_ref, instruction_ref> map_main_to_mattn;
+
+        // Add first gemm and fuse any input shape ops
+        module_ref fuse_gemm1 = mpm.create_module("reshapes_gemm1");
+        auto [anchor_op, top_inputs] =
+            fuse_input_ops_and_gemm_based_op(fuse_gemm1, gemm1->inputs(), gemm1->get_operator());
+        fuse_gemm1->add_return({anchor_op});
+        m_attn->add_params(top_inputs, &map_main_to_mattn);
+
+        std::unordered_map<instruction_ref, instruction_ref> map_gemm1_to_mattn(map_main_to_mattn);
+        auto m_gemm1 = m_attn->fuse(*fuse_gemm1, top_inputs, &map_gemm1_to_mattn).front();
+        map_main_to_mattn[gemm1] = m_gemm1;
+        mpm.remove_module("reshapes_gemm1");
+
+        // Add pointwise-softmax, unroll any pointwise modules back to base ops
+        m_attn->add_params(fused_reduce->inputs(), &map_main_to_mattn);
+        std::unordered_map<instruction_ref, instruction_ref> map_mfr_to_mattn(map_main_to_mattn);
+        auto pw_softmax =
+            m_attn
+                ->fuse(
+                    *fused_reduce->module_inputs().front(),
+                    fused_reduce->inputs(),
+                    &map_mfr_to_mattn,
+                    [&](module& main_mod,
+                        instruction_ref pos,
+                        const operation& op,
+                        const std::vector<instruction_ref>& inputs,
+                        const std::vector<module_ref>& mod_args) {
+                        if(op.name() == "pointwise")
+                        {
+                            auto* sub_pm     = mod_args.front();
+                            auto param_map_2 = create_param_map_with_literals(
+                                &main_mod, sub_pm, op.compute_shape(to_shapes(inputs), mod_args));
+                            return main_mod.insert_inline(pos, *sub_pm, inputs, &param_map_2)
+                                .front(); // cppcheck-suppress returnDanglingLifetime;
+                        }
+                        return main_mod.insert_instruction(pos, op, inputs, mod_args);
+                    })
+                .front();
+
+        // fused_reduce submodule should end with a softmax
+        auto result = match::match_instruction(*m_attn, pw_softmax, match::softmax());
+        if(result.result != pw_softmax)
         {
-            auto bias_input = orig_inputs[2];
-            instruction_ref bias_param =
-                mm->add_parameter("y_bias", bias_input->get_shape().as_standard());
-            bias = mm->add_instruction(make_op("add"), scaled_gemm0, bias_param);
-            inputs.push_back(bias_input);
+            mpm.remove_module(m_attn->name());
```

Comment:
Why do you remove the module? That should be removed automatically by DCE afterwards.

Reply:
I think I had to add this because if we hit this condition in `find_mlir_attention_fused_ops`, then we will also hit it again for `find_mlir_standalone_attention_op`. Which means it will try to create a module with the same name before DCE is run.

Reply:
Actually, I added dce between those passes, so this can be removed